File: sdk/typescript/rooch-sdk/src/transactions/transactionData.ts
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

import { sha3_256 } from '../utils/index.js'
import { normalizeRoochAddress } from '../address/index.js'
import { Args, bcs, Serializer } from '../bcs/index.js'
import { address, Bytes, identifier, u8, u64 } from '../types/index.js'
import { CallFunctionArgs, CallScript } from './types.js'

const DEFAULT_GAS = BigInt(50000000)

export class CallFunction {
  address: string
  module: identifier
  function: identifier
  args: Args[]
  typeArgs: string[]

  constructor(input: CallFunctionArgs) {
    const [pkg, mod, fn] =
      'target' in input ? input.target.split('::') : [input.address, input.module, input.function]

    this.address = pkg
    this.module = mod
    this.function = fn

    this.args = input.args || []
    this.typeArgs = input.typeArgs?.map((item) => Serializer.typeTagToString(item)) || []
  }

  functionId(): string {
    return `${normalizeRoochAddress(this.address)}::${this.module}::${this.function}`
  }

  encodeArgs(): string[] {
    return this.args?.map((item) => item.encodeWithHex())
  }

  encodeArgsWithUtf8(): string {
    return ''
  }

  encodeArgsToByteArrays(): u8[][] {
    return this.args.map((item) => item.encode()).map((item) => Array.from(item))
  }
}

type MoveActionType = CallFunction | CallScript

export class MoveAction {
  scheme: number
  val: MoveActionType

  private constructor(scheme: number, val: MoveActionType) {
    this.scheme = scheme
    this.val = val
  }

  static newCallFunction(input: CallFunctionArgs) {
    return new MoveAction(1, new CallFunction(input))
  }

  static newCallScript(input: CallScript) {
    return new MoveAction(2, input)
  }
}

export class TransactionData {
  sender?: address
  sequenceNumber?: u64
  chainId?: u64
  maxGas: u64
  action: MoveAction

  constructor(
    action: MoveAction,
    sender?: string,
    sequenceNumber?: bigint,
    chainId?: bigint,
    maxGas?: bigint,
  ) {
    this.sender = sender
    this.sequenceNumber = sequenceNumber
    this.chainId = chainId
    this.action = action
    this.maxGas = maxGas || DEFAULT_GAS
  }

  encode(): Bytes {
    const call = this.action.val as CallFunction

    return bcs.RoochTransactionData.serialize({
      sender: this.sender!,
      sequenceNumber: this.sequenceNumber!,
      chainId: this.chainId!,
      maxGas: this.maxGas,
      action: {
        kind: 'CallFunction',
        functionId: {
          moduleId: {
            address: call.address,
            name: call.module,
          },
          name: call.function,
        },
        args: Array.from(call.encodeArgsToByteArrays()),
        typeArgs: call.typeArgs,
      },
    }).toBytes()
  }

  hash(): Bytes {
    return sha3_256(this.encode())
  }
}

File: sdk/golang/rooch-sdk/transactions/transaction_data.go
// Copyright (c) RoochNetwork
// SPDX-License-Identifier: Apache-2.0

package transactions

import (
	"fmt"
	"strings"

	"github.com/rooch-network/rooch/sdk/golang/address"
	"github.com/rooch-network/rooch/sdk/golang/bcs"
	"github.com/rooch-network/rooch/sdk/golang/types"
	"github.com/rooch-network/rooch/sdk/golang/utils"
)

const DEFAULT_GAS = uint64(50000000)

type CallFunctionArgs struct {
	Target   string
	Address  string
	Module   string
	Function string
	Args     []bcs.Args
	TypeArgs []string
}

type CallFunction struct {
	Address  string
	Module   types.Identifier
	Function types.Identifier
	Args     []bcs.Args
	TypeArgs []string
}

func NewCallFunction(input CallFunctionArgs) *CallFunction {
	var pkg, mod, fn string
	if input.Target != "" {
		parts := strings.Split(input.Target, "::")
		pkg, mod, fn = parts[0], parts[1], parts[2]
	} else {
		pkg, mod, fn = input.Address, input.Module, input.Function
	}

	if input.Args == nil {
		input.Args = make([]bcs.Args, 0)
	}
	if input.TypeArgs == nil {
		input.TypeArgs = make([]string, 0)
	}

	return &CallFunction{
		Address:  pkg,
		Module:   mod,
		Function: fn,
		Args:     input.Args,
		TypeArgs: input.TypeArgs,
	}
}

func (c *CallFunction) FunctionId() string {
	return fmt.Sprintf("%s::%s::%s", 
		address.NormalizeRoochAddress(c.Address), 
		c.Module, 
		c.Function)
}

func (c *CallFunction) EncodeArgs() []string {
	result := make([]string, len(c.Args))
	for i, arg := range c.Args {
		result[i] = arg.EncodeWithHex()
	}
	return result
}

func (c *CallFunction) EncodeArgsWithUtf8() string {
	return ""
}

func (c *CallFunction) EncodeArgsToByteArrays() [][]uint8 {
	result := make([][]uint8, len(c.Args))
	for i, arg := range c.Args {
		result[i] = arg.Encode()
	}
	return result
}

type MoveActionType interface {
	isActionType()
}

func (*CallFunction) isActionType() {}
func (*CallScript) isActionType()   {}

type MoveAction struct {
	Scheme int
	Val    MoveActionType
}

func NewCallFunctionAction(input CallFunctionArgs) *MoveAction {
	return &MoveAction{
		Scheme: 1,
		Val:    NewCallFunction(input),
	}
}

func NewCallScriptAction(input *CallScript) *MoveAction {
	return &MoveAction{
		Scheme: 2,
		Val:    input,
	}
}

type TransactionData struct {
	Sender         *types.Address
	SequenceNumber *uint64
	ChainId        *uint64
	MaxGas         uint64
	Action         *MoveAction
}

func NewTransactionData(
	action *MoveAction,
	sender string,
	sequenceNumber uint64,
	chainId uint64,
	maxGas uint64,
) *TransactionData {
	if maxGas == 0 {
		maxGas = DEFAULT_GAS
	}

	var senderAddr *types.Address
	if sender != "" {
		addr := types.Address(sender)
		senderAddr = &addr
	}

	return &TransactionData{
		Sender:         senderAddr,
		SequenceNumber: &sequenceNumber,
		ChainId:        &chainId,
		MaxGas:         maxGas,
		Action:         action,
	}
}

func (t *TransactionData) Encode() []byte {
	call := t.Action.Val.(*CallFunction)

	data := bcs.RoochTransactionData{
		Sender:         *t.Sender,
		SequenceNumber: *t.SequenceNumber,
		ChainId:        *t.ChainId,
		MaxGas:         t.MaxGas,
		Action: bcs.TransactionAction{
			Kind: "CallFunction",
			FunctionId: bcs.FunctionId{
				ModuleId: bcs.ModuleId{
					Address: call.Address,
					Name:    call.Module,
				},
				Name: call.Function,
			},
			Args:     call.EncodeArgsToByteArrays(),
			TypeArgs: call.TypeArgs,
		},
	}

	return data.Serialize()
}

func (t *TransactionData) Hash() []byte {
	return utils.Sha3_256(t.Encode())
} 